File: java/src/main/java/com/github/algo/graph/RottingOranges.java
package com.github.algo.graph;

import org.javatuples.Pair;

import java.util.Deque;
import java.util.LinkedList;
import java.util.concurrent.Callable;
import java.util.function.Function;

public class RottingOranges {

    private int[][] movements = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

    public int orangesRotting(int[][] grid) {
        Deque<Pair<Integer, Integer>> queue = new LinkedList<>();
        int goodOranges = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                if (grid[i][j] == 0) grid[i][j] = -2;
                if (grid[i][j] == 2) {
                    grid[i][j] = 0;
                    queue.add(Pair.with(i, j));
                }
                if (grid[i][j] == 1) {
                    grid[i][j] = -1;
                    goodOranges++;
                }
            }
        }
        int maxMinutes = 0;
        while (!queue.isEmpty()) {
            Pair<Integer, Integer> pair = queue.removeFirst();
            int i = pair.getValue0();
            int j = pair.getValue1();
            for (int[] move : movements) {
                int di = i + move[0], dj = j + move[1];
                if (di < grid.length && di >= 0 && dj < grid[i].length && dj >= 0) {
                    if (grid[di][dj] == -1 || grid[di][dj] > grid[i][j] + 1) {
                        if (grid[di][dj] == -1) {
                            goodOranges--;
                        }
                        grid[di][dj] = grid[i][j] + 1;
                        queue.add(Pair.with(di, dj));
                        maxMinutes = Math.max(maxMinutes, grid[di][dj]);
                    }
                }
            }
        }
        return goodOranges > 0 ? -1 :maxMinutes;
    }

}

File: python/main/graphs/RottingOranges.py
from collections import deque
from typing import List

MOVEMENTS = [(1, 0), (0, 1), (-1, 0), (0, -1)]


class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        queue = deque()
        good_apples_count = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == 1:
                    grid[i][j] = -1
                    good_apples_count += 1
                elif grid[i][j] == 0:
                    grid[i][j] = -3
                elif grid[i][j] == 2:
                    grid[i][j] = 0
                    queue.append((i, j))

        maxTime = 0

        while len(queue) > 0:
            size = len(queue)
            for _ in range(size):
                i, j = queue.popleft()
                for di, dj in MOVEMENTS:
                    ni = i + di
                    nj = j + dj
                    if 0 <= ni < len(grid) and 0 <= nj < len(grid[ni]):
                        if grid[ni][nj] == -1:
                            grid[ni][nj] = grid[i][j] + 1
                            queue.append((ni, nj))
                            good_apples_count -= 1
                            maxTime = max(maxTime, grid[ni][nj])
        if good_apples_count > 0:
            return -1
        return maxTime

        pass


if __name__ == "__main__":
    solution = Solution()
    output = solution.orangesRotting([[2, 1, 1], [1, 1, 0], [0, 1, 1]])
    print(output)
